include/pybind11/stl_binders.h

```diff
@@ -0,0 +1,260 @@
+
+    Copyright (c) 2016 Sergey Lyskov
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#ifndef _INCLUDED_std_binders_h_
+#define _INCLUDED_std_binders_h_
+
+#include "common.h"
+#include "operators.h"
+
+#include <type_traits>
+#include <utility>
+#include <algorithm>
+#include <sstream>
+
+
+NAMESPACE_BEGIN(pybind11)
+
+#if (!defined _MSC_VER) || (_MSC_VER > 1900)
+	#define INSERTION_OPERATOR_IMPLEMENTATION
+#endif
+
+
+template<typename T>
+constexpr auto has_equal_operator(int) -> decltype(std::declval<T>() == std::declval<T>(), bool()) { return true; }
+template<typename T>
+constexpr bool has_equal_operator(...) { return false; }
+
+
+
+template<typename T>
+constexpr auto has_not_equal_operator(int) -> decltype(std::declval<T>() != std::declval<T>(), bool()) { return true; }
+template<typename T>
+constexpr bool has_not_equal_operator(...) { return false; }
+
+
+#ifdef INSERTION_OPERATOR_IMPLEMENTATION
+namespace has_insertion_operator_implementation {
+enum class False {};
+struct any_type {
+    template<typename T> any_type(T const&);
+};
+False operator<<(std::ostream const&, any_type const&);
+}
+template<typename T>
+constexpr bool has_insertion_operator() {
+	using namespace has_insertion_operator_implementation;
+	return std::is_same< decltype(std::declval<std::ostream&>() << std::declval<T>()), std::ostream & >::value;
+}
+#endif
+
+
+template <typename T, typename Allocator = std::allocator<T>, typename holder_type = std::unique_ptr< std::vector<T, Allocator> > >
+class vector_binder {
+	using Vector = std::vector<T, Allocator>;
+	using SizeType = typename Vector::size_type;
+
+	using Class_ = pybind11::class_<Vector, holder_type >;
+
+	// template<typename U = T, typename std::enable_if< std::is_constructible<U>{} >::type * = nullptr>
+	// void maybe_constructible(Class_ &cl) {
+	// 	cl.def(pybind11::init<>());
+	// }
+	// template<typename U = T, typename std::enable_if< !std::is_constructible<U>{} >::type * = nullptr>
+	// void maybe_constructible(Class_ &cl) {}
+
+	template<typename U = T, typename std::enable_if< std::is_default_constructible<U>{} >::type * = nullptr>
+	void maybe_default_constructible() {
+		cl.def(pybind11::init<SizeType>());
+		cl.def("resize", (void (Vector::*)(SizeType count)) &Vector::resize, "changes the number of elements stored");
+
+		/// Slicing protocol
+		cl.def("__getitem__", [](Vector const &v, pybind11::slice slice) -> Vector * {
+				pybind11::ssize_t start, stop, step, slicelength;
+				if(!slice.compute(v.size(), &start, &stop, &step, &slicelength))
+					throw pybind11::error_already_set();
+				Vector *seq = new Vector(slicelength);
+				for (int i=0; i<slicelength; ++i) {
+					(*seq)[i] = v[start]; start += step;
+				}
+				return seq;
+			});
+	}
+	template<typename U = T, typename std::enable_if< !std::is_default_constructible<U>{} >::type * = nullptr>
```

Comment:
I have not seen this syntax before (``std::enable_if< !std::is_default_constructible<U>{} >::type``) and use the following elsewhere in the codebase: ``std::enable_if< !std::is_default_constructible<U>::value>::type``. Just mentioning it in case it makes any difference for MSVC